### ISSUE TITLE
Help wanted with text generator code

### ISSUE THREAD

**Issue by tararengan** [tararengan/NLP#2]
Codebase:
https://github.com/tararengan/NLP/blob/dfd00d146ccfc8a62fd2d012954e03d502270b7a/textgenerator.py#L1-L253

I tried to write code to generate text from sample text using first principles. It is based on Andrej Karpathy's code: <script src="https://gist.github.com/karpathy/d4dee566867f8291f086.js"></script>

 The structure of the network is simple. There are two layers.

1. hidden = tanh(W_xh * x + W_hh*hidden_prev + b)
2. y = W_hy*hidden + b_prime

where x is one-hot encoded vector corresponding to input character (a.k.a data point).

Finally, there is an exponentiation followed by weight normalization on the elements of y to get probabilities. The classes are the letters in the vocabulary of the training set and the loss function is cross entropy.

I wanted to accomplish the following:

- No libraries, just plain python and numpy. 
- Since the first layer takes in as (one of the) input the hidden vector from forward pass on the previous data point, there is a recursive relationship inherent in the system. Specifically, the gradient of hidden w.r.t the variables W_xh, W_hh and b comes from two contributions - one directly from the formula above in (a)treating x and h as constants, and (b)chain rule on hidden_prev. To deal with the second one,   instead of doing a reverse iteration as Karpathy does to get gradients, I decided to try forward iteration. The idea is to iterate on the sequence on data points **as they appear in the training set**. Setting the initial hidden_prev to a vector of ones, one can get the gradients of hidden of the current iteration w.r.t the variables W_xh, W_hh and b. Then, this is passed on to the next iteration which uses these values in the chain rule for the contribution (b). Anything wrong here?

I implement above in the following manner:
There are 10 epochs. In each, the input text is divided into batches of preset size. Each batch defines the value of the loss function across the data points in it, i.e. the characters that appear in the snippet corresponding to the batch. Thus, in one single batch, there is a forward pass and gradient value, loss function value update as well as storage of hidden layer gradients(my idea above) with each data point. When the batch is done, the gradient descent step is taken along the direction of the negative gradient. Then using these new parameter values (parameters being the W's and b's), the next batch is considered and so on. When all batches are done, the resulting parameter values act as starting point for the next epoch. Sound right?

The code doesn't work. It spits out nonsensical text that admittedly comes from the input vocab but that has very little to do with the input text itself unless the latter is something absurdly simplistic. The loss function value does not decrease with each batch. 

 The error has to be in one or more of the following:

(A) differentiation is flawed w.r.t system architecture as captured in forward pass
(B) sampling text from the model after training is erroneous
(C)parameters for training are set poorly

This is my first time with RNN and text generation (have used RNN itself before for classifying points in the real n-dimensional space).

**Comment by tararengan** [comment]
Solved. Learning rate was too conservative, previous value of  hidden and derivates must be passed to new batches,  and turning vectors rtf arrays sped up code.
